tentacle/tests/test_tls_dial.rs

```diff
@@ -0,0 +1,432 @@
+    // Specially for server.crt not a cert-chain only one server certificate, so manually make
+    // a cert-chain.
+    if certs.len() == 1 && !cacerts.is_empty() {
+        certs.extend(cacerts);
+    }
+
+    server_config
+        .set_single_cert_with_ocsp_and_sct(certs, privkey, vec![], vec![])
+        .expect("bad certificates/private key");
+
+    if config.cypher_suits.is_some() {
+        server_config.ciphersuites = lookup_suites(
+            &config
+                .cypher_suits
+                .as_ref()
+                .expect("cypher_suits option error"),
+        );
+    }
+
+    if config.protocols.is_some() {
+        server_config.versions = lookup_versions(config.protocols.as_ref().unwrap());
+        server_config.set_protocols(
+            &config
+                .protocols
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|proto| proto.as_bytes().to_vec())
+                .collect::<Vec<_>>()[..],
+        );
+    }
+
+    server_config
+}
+
+/// Build a `ClientConfig` from our NetConfig
+pub fn make_client_config(config: &NetConfig) -> ClientConfig {
+    let mut client_config = ClientConfig::new();
+    client_config.key_log = Arc::new(KeyLogFile::new());
+
+    if config.cypher_suits.is_some() {
+        client_config.ciphersuites = lookup_suites(config.cypher_suits.as_ref().unwrap());
+    }
+
+    if config.protocols.is_some() {
+        client_config.versions = lookup_versions(config.protocols.as_ref().unwrap());
+
+        client_config.set_protocols(
+            &config
+                .protocols
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|proto| proto.as_bytes().to_vec())
+                .collect::<Vec<_>>()[..],
+        );
+    }
+
+    let cafile = config.ca_cert.as_ref().unwrap();
+
+    let certfile = fs::File::open(cafile).expect("Cannot open CA file");
+    let mut reader = BufReader::new(certfile);
+    client_config.root_store.add_pem_file(&mut reader).unwrap();
+
+    if config.server_key.is_some() || config.server_cert_chain.is_some() {
+        load_key_and_cert(
+            &mut client_config,
+            config
+                .server_key
+                .as_ref()
+                .expect("must provide client_key with client_cert"),
+            config
+                .server_cert_chain
+                .as_ref()
+                .expect("must provide client_cert with client_key"),
+            cafile,
+        );
+    }
+
+    client_config
+}
+
+fn test_tls_dial() {
+    let (meta_1, receiver_1) = create_meta(1.into());
+    let (meta_2, receiver_2) = create_meta(1.into());
+    let (shandle, _error_receiver_1) = create_shandle(true);
+    let (addr_sender, addr_receiver) = channel::oneshot::channel::<Multiaddr>();
+
+    thread::spawn(move || {
+        let multi_addr_1 = Multiaddr::from_str(
+            "/ip4/127.0.0.1/tcp/0/tls/0x09cbaa785348dabd54c61f5f9964474f7bfad7df",
```

Comment:
I'm confused, I thought tls was followed by a domain name, for example: `/ip4/127.0.0.1/tcp/0/tls/test.example.com` but here it looks like a hex string, could you explain the detail, thanks.

Reply:
hex string also can be used as "domain name". It just a check with certificate, and the hex string need equal with the `common name` of `certificates/node0/server.crt `. And the client need know it for tls handshake, but server didn't. So we say tls for client and server is not equal.
What's more change the hex string would make the test never stop, 😅. May be need modify?

Reply:
understood, thanks.